### ISSUE TITLE
Make parse errors in statements recoverable

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#581]
These should recover by looking for the next ";"

**Comment by matozoid** [comment]
We could insert a "BrokenStatement" in the AST with the literal text of the statement in it.

**Comment by martinlippert** [comment]
I am very interested in this whole "make javaparser recoverable from errors" story. Any news on this?

**Comment by matozoid** [comment]
It's not currently being worked on, but the steps to take are more or less clear.
1. We need to play around with the information on this page: https://javacc.java.net/doc/errorrecovery.html to make sure the parser doesn't throw parse errors, but only inserts a "bad node" and adds a Problem. 
2. We need to move the semantic checks (meaning: the check in Java code) in the grammar to a validation framework that we yet have to build.

There are issues for both of these, waiting for someone to pick them up :-)

**Comment by martinlippert** [comment]
Thanks for the details. At the moment we don't have the cycles to work on this on our own. If that changes (and this isn't already done), I will let you know for sure.

**Comment by ftomassetti** [comment]
_Here is the marketing department speaking_ I could not help but notice that your GitHub profile says "Pivotal". It would be could for us to brag about Pivotal being using JavaParser. We have to start creating our testimonials page. Do you think would be possible to bother some of you with a couple of questions on JavaParser? It would not take long, you should just say "it is great, it changed my life and my dog loves me more since I use JavaParser". That would mean a lot to us. _End of marketing comunication, back to business_

**Comment by martinlippert** [comment]
Hey _marketing department_, I understand and support your request, but it is way too early for us to provide a quote. We need a gather a lot more experience using this before we can be part of a testimonials page and haven't even decided yet which exact way to go. I would prefer to use this project for parsing code, but since we are working on code editing capabilities, the error recovery is a very essential part. So this missing piece is preventing us from diving deeper into this at the moment, but again, we are at the very beginning here. But we can for sure come back to the _marketing department_ questions then.

**Comment by ftomassetti** [comment]
thanks @martinlippert and for anything feel free to reach us

**Comment by matozoid** [comment]
@jeandersonbc is now working on this.

**Comment by martinlippert** [comment]
great news and let me know if there is something to try

**Comment by jeandersonbc** [comment]
glad to hear you are interested @martinlippert! we are thinking on make something based on the so-called "panic mode" recovery, which basically skips the input stream until it finds a "synchronization token" (like a semicolon or closing brackets) to continue the parsing:

```{java}
import; // first parsing error

class Foo {
     int bar() {
         return 3+; // second parsing error
     }
}
```

The downside is that this approach depends on the presence of the synchronization token:

```{java}
import  // bad import

class Foo {
     int bar() {
         return 3+; // the parse would only recover at this point
     }
}
```

@matozoid just for clarification, by "statements" you mean every production ended by a semicolon (eg, import, package, variable declaration, etc) and not productions derived from `Statement()`, right? (especially because `Block` is a kind of statement but the synchronization token would be "}")

**Comment by matozoid** [comment]
@jeandersonbc my plan was to do this step by step, so first only `Statement()`s. That way you don't have to work for weeks on a big blob. Also nicer for reviewers :) Feel free to use your own judgement.

**Comment by jeandersonbc** [comment]
@matozoid baby steps then! I noticed that already exists an issue for statements surrounded by `{` and `}` (see #582) so I will skip any of the statements that fit on #852

**Comment by matozoid** [comment]
Feel free to open extra issues or do them for this issue anyway :)
